From 566ccf3d252458d0a1236fefcb4b0389add8358f Mon Sep 17 00:00:00 2001
From: Allen Wild <allenwild93@gmail.com>
Date: Sat, 20 Jan 2018 01:44:55 -0500
Subject: [PATCH] ncurses: make ABI version user-selectable

The ncurses6 features can be useful and backwards-compatibility
with the ncurses5 ABI isn't always needed.

Remove the narrowc libtinfo libraries so that widec versions get
installed to prevent runtime segfaults. Not sure why this is only needed
for ABI v6.

Widec is the default now, so to build narrowc ncurses we need to
explicitly pass --disable-widec.

Signed-off-by: Allen Wild <allenwild93@gmail.com>
---
 meta/recipes-core/ncurses/ncurses.inc    | 15 +++++++++++----
 meta/recipes-core/ncurses/ncurses_6.5.bb |  2 +-
 2 files changed, 12 insertions(+), 5 deletions(-)

diff --git a/meta/recipes-core/ncurses/ncurses.inc b/meta/recipes-core/ncurses/ncurses.inc
index f5e37b94da..1b42e739cb 100644
--- a/meta/recipes-core/ncurses/ncurses.inc
+++ b/meta/recipes-core/ncurses/ncurses.inc
@@ -20,6 +20,9 @@ EXTRA_AUTORECONF = "-I m4"
 CACHED_CONFIGUREVARS = "cf_cv_func_nanosleep=yes"
 CACHED_CONFIGUREVARS:append:linux = " cf_cv_working_poll=yes"
 
+# ncurses ABI version: 5 for compatibility with most external
+# applications, 6 for new features such as extra mouse buttons
+NCURSES_ABI_VERSION ?= "5"
 # Whether to enable separate widec libraries; must be 'true' or 'false'
 #
 # TODO: remove this variable when widec is supported in every setup?
@@ -101,7 +104,7 @@ do_configure() {
 	# The --enable-pc-files requires PKG_CONFIG_LIBDIR existed
 	mkdir -p ${PKG_CONFIG_LIBDIR}
 	( cd ${S}; gnu-configize --force )
-	ncurses_configure "narrowc" || \
+	ncurses_configure "narrowc" "--disable-widec" || \
 		return 1
 	! ${ENABLE_WIDEC} || \
 		ncurses_configure "widec" "--enable-widec" "--without-progs"
@@ -172,8 +175,12 @@ do_install() {
                 install.data
 
 
-        ! ${ENABLE_WIDEC} || \
+        if ${ENABLE_WIDEC}; then
+            # remove libtinfo files to force the widec versions to get installed, otherwise
+            # segfaults happen at runtime
+            rm ${D}${libdir}/libtinfo*
             oe_runmake -C widec ${_install_cfgs} ${_install_opts}
+        fi
 
         cd narrowc
 
@@ -212,7 +219,7 @@ do_install() {
                 test -h $f || continue
                 rm -f $f
                 echo '/* GNU ld script */'  >$f
-                echo "INPUT($i.so.5 AS_NEEDED(-ltinfo))" >>$f
+                echo "INPUT($i.so.${NCURSES_ABI_VERSION} AS_NEEDED(-ltinfo))" >>$f
         done
 
         # Make sure that libcurses is linked so that it gets -ltinfo
@@ -237,7 +244,7 @@ do_install() {
 
             # Use ln -rs to ensure this is a relative link despite absolute paths
             # (as we can't know the relationship between base_libdir and libdir).
-            ln -rs ${D}${base_libdir}/libtinfo.so.5 ${D}${libdir}/libtinfo.so
+            ln -rs ${D}${base_libdir}/libtinfo.so.${NCURSES_ABI_VERSION} ${D}${libdir}/libtinfo.so
         fi
         if [ -d "${D}${includedir}/ncurses" ]; then
             for f in `find ${D}${includedir}/ncurses -name "*.h"`
diff --git a/meta/recipes-core/ncurses/ncurses_6.5.bb b/meta/recipes-core/ncurses/ncurses_6.5.bb
index 2e3ee337ea..f9d06ea6ed 100644
--- a/meta/recipes-core/ncurses/ncurses_6.5.bb
+++ b/meta/recipes-core/ncurses/ncurses_6.5.bb
@@ -8,7 +8,7 @@ SRC_URI += "file://0001-tic-hang.patch \
 # commit id corresponds to the revision in package version
 SRCREV = "1c55d64d9d3e00399a21f04e9cac1e472ab5f70a"
 S = "${WORKDIR}/git"
-EXTRA_OECONF += "--with-abi-version=5"
+EXTRA_OECONF += "--with-abi-version=${NCURSES_ABI_VERSION}"
 UPSTREAM_CHECK_GITTAGREGEX = "v(?P<pver>\d+_\d+)$"
 
 # This is needed when using patchlevel versions like 6.1+20181013
-- 
2.48.1

